# Исследование рынка общественного питания Москвы

Вы решили открыть небольшое кафе в Москве. Оно оригинальное — гостей должны обслуживать роботы. Проект многообещающий, но дорогой. 

Вместе с партнёрами вы решились обратиться к инвесторам. Их интересует текущее положение дел на рынке — сможете ли вы снискать популярность на долгое время, когда все зеваки насмотрятся на роботов-официантов?

**Цель:** Исследовать рынок общественного питания Москвы и дать рекомендации о виде заведения, количество посадочных мест и районе расположения.

Описание данных:
* id — идентификатор объекта;
* object_name — название объекта общественного питания;
* chain — сетевой ресторан;
* object_type — тип объекта общественного питания;
* address — адрес;
* number — количество посадочных мест.


## Изучение данных из файла

Установим модуль Dadata для стандартизации адресов

In [1]:
!pip install dadata -U

     |████████████████████████████████| 76 kB 3.6 MB/s 
     |████████████████████████████████| 58 kB 6.3 MB/s 
     |████████████████████████████████| 77 kB 6.1 MB/s 
     |████████████████████████████████| 54 kB 3.4 MB/s 


И еще folium для геопространственной визуализации

In [2]:
!pip install folium -U

     |████████████████████████████████| 94 kB 2.9 MB/s 
  Attempting uninstall: folium
    Found existing installation: folium 0.8.3
    Uninstalling folium-0.8.3:
      Successfully uninstalled folium-0.8.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
datascience 0.10.6 requires folium==0.2.1, but you have folium 0.12.1 which is incompatible.


И обновим плотли

In [3]:
!pip install plotly -U

     |████████████████████████████████| 20.6 MB 85.3 MB/s 
  Attempting uninstall: plotly
    Found existing installation: plotly 4.4.1
    Uninstalling plotly-4.4.1:
      Successfully uninstalled plotly-4.4.1


In [4]:
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px
import numpy as np
import re
import folium
import folium.plugins as plugins
import requests

from dadata import Dadata
from tqdm.notebook import tqdm

pd.set_option('chained_assignment', None)

In [5]:
try:
    rest_data = pd.read_csv('https://code.s3.yandex.net/datasets/rest_data.csv')
except Exception as err:
    print('Ошибка:', err)

In [6]:
rest_data.head()

id  ... number
0  151635  ...     48
1   77874  ...     35
2   24309  ...     95
3   21894  ...     40
4  119365  ...     50

[5 rows x 6 columns]

## Предобработка данных

In [7]:
def primary_info(df):
    display(df.head())
    print()
    print('Проверка на пропущенные значения:')
    display(pd.DataFrame(round((df.isna().mean()*100),2)).style.background_gradient('coolwarm'))
    print('----------------')
    print()
    print('Проверка типа данных на корректность:')
    print(df.info())
    print('----------------')
    print()
    print('Количество дубликатов:', df.duplicated().sum())

In [8]:
primary_info(rest_data)

id  ... number
0  151635  ...     48
1   77874  ...     35
2   24309  ...     95
3   21894  ...     40
4  119365  ...     50

[5 rows x 6 columns]


Проверка на пропущенные значения:


----------------

Проверка типа данных на корректность:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15366 entries, 0 to 15365
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   id           15366 non-null  int64 
 1   object_name  15366 non-null  object
 2   chain        15366 non-null  object
 3   object_type  15366 non-null  object
 4   address      15366 non-null  object
 5   number       15366 non-null  int64 
dtypes: int64(2), object(4)
memory usage: 720.4+ KB
None
----------------

Количество дубликатов: 0


Полных дубликатов нет, проверим датафрейм без столбца id

In [9]:
rest_data[rest_data.loc[:, ('object_name', 'chain', 'object_type', 'address', 'number')].duplicated()]

id  ... number
2381    28611  ...    100
3481    20914  ...     40
4431    23759  ...    100
4537    23391  ...     15
5059    20021  ...     20
...       ...  ...    ...
14578  222932  ...    120
14596  204947  ...      0
14844  219518  ...      0
15173  218542  ...      0
15217  214146  ...      0

[82 rows x 6 columns]

Вот и они - 82 дубликата. Удалим их

In [10]:
rest_data.drop_duplicates(subset=['object_name', 'chain', 'object_type', 'address', 'number'], inplace=True)

In [11]:
rest_data[rest_data.loc[:, ('object_name', 'chain', 'object_type', 'address')].duplicated()]

id  ... number
115     28614  ...     12
580    149205  ...     35
581    149131  ...     39
1409   151267  ...     45
2375    28604  ...    150
...       ...  ...    ...
15006  219563  ...     25
15007  219569  ...      2
15216  214468  ...      0
15363  209186  ...     20
15365  222535  ...     10

[88 rows x 6 columns]

Если убрать столбец с количеством посадочных мест, то появятся еще 88 дубликатов. Но ввиду недостаточной информации, их обработать не получится. Это могут быть как реальные дубликаты, к примеру, ресторан увеличил/уменьшил количество посадочных мест, добавили новую информацию, а старую запись удалить забыли, так и ресторан с несколькими залами - к примеру, банкетный и гостевой залы с разным количеством посадочных мест.

Приведем все названия заведений к нижнему регистру, сохранив заглавной первую букву

In [12]:
rest_data['object_name'] = rest_data['object_name'].str.capitalize()

In [13]:
rest_data

id  ... number
0      151635  ...     48
1       77874  ...     35
2       24309  ...     95
3       21894  ...     40
4      119365  ...     50
...       ...  ...    ...
15361  208537  ...     50
15362  209264  ...     10
15363  209186  ...     20
15364  221900  ...     36
15365  222535  ...     10

[15284 rows x 6 columns]

**Вывод**
* Пропущенных значений нет
* С типами данных все в порядке
* Удалили 82 дубликата
* Исправили регистр в названиях заведений

Можно приступать к анализу

## Анализ данных

### Исследуем соотношение видов объектов общественного питания по количеству.


In [14]:
object_type_count = rest_data['object_type'].value_counts().reset_index()
object_type_count.columns = ['object_type', 'count']
object_type_count

object_type  count
0                               кафе   6071
1                           столовая   2584
2                           ресторан   2282
3  предприятие быстрого обслуживания   1898
4                                бар    855
5                              буфет    577
6                          кафетерий    395
7                         закусочная    349
8          магазин (отдел кулинарии)    273

In [15]:
fig = px.bar(object_type_count,
             x='count', y='object_type', orientation='h',
             title='Соотношение видов объектов общественного питания по количеству')

fig.show()

In [16]:
fig = px.pie(object_type_count, values='count', names='object_type', 
             title='Соотношение видов объектов общественного питания по количеству')

fig.show()

* Абсолютный лидер в этой категории - кафе. Почти 40% (6099) всех заведений общественного питания приходится на него;
* Далее по популярности - столовые (16.8%) и рестораны (14.9%);
* Наименее популярны по количеству заведений - это отделы кулинарии (1.78%) и закусочные (2.34%).

### Исследуем соотношение сетевых и несетевых заведений по количеству.


In [17]:
chain_count = rest_data['chain'].value_counts().reset_index()
chain_count.columns = ['chain', 'count']
chain_count

chain  count
0   нет  12320
1    да   2964

In [18]:
fig = px.pie(chain_count, values='count', names='chain', 
             title='Соотношение сетевых и несетевых заведений по количеству')

fig.update_layout(
    legend_title_text='Сетевое заведение'
)
fig.show()

Подавляющее большинство заведений (80%) - несетевые.

### Для какого вида объекта общественного питания характерно сетевое распространение?

In [ ]:
object_type_chain_count = (
    rest_data
        .query('chain == "да"')
        .groupby('object_type', as_index=False)
        .agg({'id':'count'})
        .rename(columns={'id':'count'})
        .sort_values(by='count')
)

object_type_chain_count

object_type  count
8                           столовая      3
1                              буфет     11
0                                бар     37
4                          кафетерий     52
2                         закусочная     56
5          магазин (отдел кулинарии)     78
7                           ресторан    543
6  предприятие быстрого обслуживания    788
3                               кафе   1396

In [ ]:
fig = px.bar(object_type_chain_count,
             x='count', y='object_type', orientation='h',
             title='Сетевые заведения по видам объектов общественного питания')

fig.show()

Больше всего сетевых заведений у кафе, но в данном случае, я считаю, лучше посчитать не в абсолютных показателях, а в относительных

In [ ]:
object_type_chain_rate = pd.merge(object_type_chain_count, object_type_count, on='object_type')
object_type_chain_rate.rename(columns={'count_x':'chain_count',
                                       'count_y':'total_count'}, inplace=True)

object_type_chain_rate['rate'] = object_type_chain_rate['chain_count'] / object_type_chain_rate['total_count']
object_type_chain_rate

object_type  chain_count  total_count      rate
0                           столовая            3         2584  0.001161
1                              буфет           11          577  0.019064
2                                бар           37          855  0.043275
3                          кафетерий           52          395  0.131646
4                         закусочная           56          349  0.160458
5          магазин (отдел кулинарии)           78          273  0.285714
6                           ресторан          543         2282  0.237949
7  предприятие быстрого обслуживания          788         1898  0.415174
8                               кафе         1396         6071  0.229946

In [ ]:
fig = px.bar(object_type_chain_rate.sort_values(by='rate'),
             x='rate', y='object_type', orientation='h',
             title='Сетевые заведения по видам объектов общественного питания')

fig.show()

Ну вот, выходит для кафе вовсе не характерно сетевое распространение.

А лидер здесь - предприятия быстрого обслуживания. Почти 42% из всех организаций - сетевые. Оно и логично - такие заведения относительно легко масштабируются и пользуются спросом для франчайзинга.

### Что характерно для сетевых заведений: много заведений с небольшим числом посадочных мест в каждом или мало заведений с большим количеством посадочных мест?


In [ ]:
df = (
    rest_data
        .query('chain == "да"')
        .groupby('object_name')
        .agg({'id':'count',
              'number':'mean'})
        .rename(columns={'id':'object_count'})
)

df

object_count      number
object_name                                            
Beverly hills diner                       1   88.000000
Bierloga                                  1   75.000000
Black & white                             1   40.000000
Bocconcino                                3   66.666667
Boobo                                     1   46.000000
...                                     ...         ...
Якитория и ян примус                      1  282.000000
Ямми микс                                 1    2.000000
Ян примус                                 1  300.000000
Японский ресторан «ваби-саби»             1  155.000000
Япоша                                     2  105.000000

[577 rows x 2 columns]

In [ ]:
fig = px.scatter(df, x='object_count', y='number',
                title='Совместное распределение количества посадочных мест и количества сетевых заведений')

fig.update_layout(
    xaxis_title='Количество заведений в сети',
    yaxis_title='Количество посадочных мест',
    xaxis_range=[0,9],
    yaxis_range=[0, (df['number'].quantile(.75) - df['number'].quantile(.25)) * 3]
)

fig.update_traces(
    marker=dict(
            color='rgba(135, 206, 250, 0.5)',
            line=dict(
                color='MediumPurple',
                width=1
            )
        )
)

fig.add_hline(y=df['number'].mean(), line_width=0.7, line_color='grey')
fig.add_vline(x=df['object_count'].mean(), line_width=0.7, line_color='grey')

fig.add_annotation(x=7.5, y=df['number'].mean(),
            text='Среднее число посадочных мест',
            showarrow=True,
            arrowhead=1)

fig.add_annotation(x=df['object_count'].mean(), y=120,
            text='Среднее число заведений в сети',
            showarrow=True,
            arrowhead=1)

fig.show()

Выходит, если взять случайное сетевое заведение, то вероятнее всего это будет сеть с малым числом заведений в сети и небольшим числом посадочных мест.

### Для каждого вида объекта общественного питания опишем среднее количество посадочных мест. Какой вид предоставляет в среднем самое большое количество посадочных мест?


In [ ]:
fig = px.box(rest_data, x='number', y='object_type', orientation='h',
             title='Диаграмма размаха по видам объектов общественного питания')

fig.update_layout(
    xaxis_range=[-5,460]
)
fig.show()

В среднем самое большое количество посадочных мест в столовых - 103 места. На 2 месте рестораны - 80 мест.

### Выделим в отдельный столбец информацию об улице из столбца address.

Это можно легко сделать с помощью Dadata или Yandex decoder'а, но они платные, поэтому сделаю через регулярные выражения

In [ ]:
words = ['улица','ул', 'переулок','шоссе','проспект','площадь','проезд',
         'село','аллея','бульвар','набережная','тупик','линия', 'микрорайон',
         'квартал', 'деревня', 'посёлок', 'километр', 'территория', 'корпус']

str_pat = r".*,\s*\b([^,]*?(?:{})\b[^,]*)[,$]+".format("|".join(words))

rest_data['street'] = rest_data['address'].str.extract(str_pat, flags=re.I)
rest_data

id             object_name  ... number                  street
0      151635                 Сметана  ...     48   улица Егора Абакумова
1       77874                  Родник  ...     35        улица Талалихина
2       24309         Кафе «академия»  ...     95    Абельмановская улица
3       21894              Пиццетория  ...     40      Абрамцевская улица
4      119365  Кафе «вишневая метель»  ...     50      Абрамцевская улица
...       ...                     ...  ...    ...                     ...
15361  208537             Шоколадница  ...     50  3-й Крутицкий переулок
15362  209264             Шоколадница  ...     10      улица Земляной Вал
15363  209186             Шоколадница  ...     20      улица Земляной Вал
15364  221900             Шоколадница  ...     36           22-й километр
15365  222535             Шоколадница  ...     10       Ходынский бульвар

[15284 rows x 7 columns]

Проверим пропуски

In [ ]:
rest_data.query('street != street')

id               object_name  ... number street
859     23601     Буфет - гбоу сош № 64  ...    100    NaN
860     23604           Буфет - гбоу 60  ...    100    NaN
1405    20199                       Это  ...     35    NaN
1406    68113                    Френдс  ...     34    NaN
1407    20105  Кафе «граф монте-кристо»  ...     40    NaN
...       ...                       ...  ...    ...    ...
14468  205933   Кафе «империя бургеров»  ...     24    NaN
14470  220595               Пекарня № 1  ...      6    NaN
14473  220978            Бар суши takai  ...     28    NaN
14478  211676                Никольский  ...    120    NaN
15184  220587               Пицца паоло  ...      0    NaN

[188 rows x 7 columns]

Осталось 188 пропусков, это 1% от наших данных. Думаю, на дальнейшее исследование это не должно повлиять

In [ ]:
rest_data = rest_data.query('street == street')
rest_data

id             object_name  ... number                  street
0      151635                 Сметана  ...     48   улица Егора Абакумова
1       77874                  Родник  ...     35        улица Талалихина
2       24309         Кафе «академия»  ...     95    Абельмановская улица
3       21894              Пиццетория  ...     40      Абрамцевская улица
4      119365  Кафе «вишневая метель»  ...     50      Абрамцевская улица
...       ...                     ...  ...    ...                     ...
15361  208537             Шоколадница  ...     50  3-й Крутицкий переулок
15362  209264             Шоколадница  ...     10      улица Земляной Вал
15363  209186             Шоколадница  ...     20      улица Земляной Вал
15364  221900             Шоколадница  ...     36           22-й километр
15365  222535             Шоколадница  ...     10       Ходынский бульвар

[15096 rows x 7 columns]

### Построим график топ-10 улиц по количеству объектов общественного питания. Воспользуемся внешней информацией для ответа на вопрос — в каких районах Москвы находятся эти улицы?


In [ ]:
top10_street = (
    rest_data
        .groupby('street', as_index=False)
        .agg({'id':'count'})
        .rename(columns={'id':'count'})
        .sort_values(by='count', ascending=False)
        .head(10)
)

top10_street

street  count
1517           проспект Мира    204
1044       Профсоюзная улица    178
719   Ленинградский проспект    170
1029  Пресненская набережная    167
434         Варшавское шоссе    160
722       Ленинский проспект    146
1513    проспект Вернадского    126
709     Кутузовский проспект    114
639     Кировоградская улица    107
1313       Ходынский бульвар    102

In [ ]:
fig = px.bar(top10_street,
             x='count', y='street', orientation='h',
             title='Топ-10 улиц по количеству объектов общественного питания')

fig.show()

А вот тут уже можно обратиться к Dadata. Функция возвращает нам район и координаты по улице

**Функция может не сработать, потому что лимит на количество бесплатных запросов. Поэтому отсюда запускать ничего не надо (да и не получится, токена нет)**

**Я сохраню полученный мною df, выложу его на файлообменник и прочитаю ниже**

In [ ]:
token = 'api'
secret = 'secret-key'
dadata = Dadata(token, secret)

In [ ]:
def get_district_dadata(row):
    query = dadata.clean("address", row['street'])
  
    try:
        return pd.Series([query['city_district_with_type'], float(query['geo_lon']), float(query['geo_lat'])])
    except:
        return pd.Series([0, 0, 0])

In [ ]:
tqdm.pandas()

top10_street[['district', 'longitude', 'latitude']] = top10_street.progress_apply(get_district_dadata, axis=1)

In [ ]:
top10_street.to_csv('top10_street.csv', encoding='utf-8', index=False)

**Отсюда можно запускать**

In [ ]:
url = 'https://drive.google.com/file/d/1no_ek_Yj0JNGXhdDss6oMVP6jLoUz8s5/view?usp=sharing'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
top10_street = pd.read_csv(path)

In [ ]:
top10_street

street  count  ...  longitude   latitude
0           проспект Мира    204  ...  37.637937  55.812368
1       Профсоюзная улица    178  ...  37.532511  55.649520
2  Ленинградский проспект    170  ...  37.545604  55.794196
3  Пресненская набережная    167  ...  37.541949  55.747144
4        Варшавское шоссе    160  ...  37.604053  55.599992
5      Ленинский проспект    146  ...  37.537038  55.682601
6    проспект Вернадского    126  ...  37.515248  55.681664
7    Кутузовский проспект    114  ...  37.509035  55.735108
8    Кировоградская улица    107  ...  37.604962  55.614296
9       Ходынский бульвар    102  ...  37.535248  55.785717

[10 rows x 5 columns]

In [ ]:
top10_street_list = list(top10_street['street'])
top10_street_list

['проспект Мира',
 'Профсоюзная улица',
 'Ленинградский проспект',
 'Пресненская набережная',
 'Варшавское шоссе',
 'Ленинский проспект',
 'проспект Вернадского',
 'Кутузовский проспект',
 'Кировоградская улица',
 'Ходынский бульвар']

In [ ]:
top10_address = (
    rest_data
        .query('street in @top10_street_list')['address']
        .value_counts()
        .reset_index()
        .rename(columns={'index':'street',
                         'address':'count'})
)

top10_address

street  count
0               город Москва, Ходынский бульвар, дом 4     95
1          город Москва, Пресненская набережная, дом 2     63
2       город Москва, проспект Мира, дом 211, корпус 2     60
3          город Москва, Кировоградская улица, дом 13А     52
4           город Москва, Кутузовский проспект, дом 57     40
..                                                 ...    ...
520  город Москва, Ленинский проспект, дом 42, корп...      1
521  город Москва, Профсоюзная улица, дом 24, корпус 1      1
522    город Москва, проспект Мира, дом 68, строение 3      1
523            город Москва, Профсоюзная улица, дом 69      1
524           город Москва, Варшавское шоссе, дом 132А      1

[525 rows x 2 columns]

**Здесь всё так же, как сверху. Запускать не надо, я дальше все сохранил.**

Строк много, Dadata исчерпчал свой лимит, будем теперь использовать Yandex Geocoder.

Функция заменяет пробелы в названиях улиц на '+', вставляет в geocode url, далее получает json, парсит, вытаскивает координаты, преобразует в float, возвращает series с координатами, а мы их сохраням в соответствующие столбцы.

Используем конструкцию try-except на случай, если url по улице не откроется

In [ ]:
API = 'api-key'

In [ ]:
def get_pos_yandexgeo(row):
    street = '+'.join(row['street'].split())

    url = 'https://geocode-maps.yandex.ru/1.x'
    params = {'geocode':'Москва,{}'.format(street),
              'format':'json',
              'apikey': API}

    r = requests.get(url, params=params).json()
    
    try:
        pos = r['response']['GeoObjectCollection']['featureMember'][0]['GeoObject']['Point']['pos']
        return pd.Series([float(pos.split()[0]), float(pos.split()[1])])
    except:
        return pd.Series([0, 0])

In [ ]:
tqdm.pandas()

top10_address[['longitude', 'latitude']] = top10_address.progress_apply(get_pos_yandexgeo, axis=1)

In [ ]:
top10_address.to_csv('top10_address.csv', encoding='utf-8', index=False)

**отсюда**

In [ ]:
url = 'https://drive.google.com/file/d/1ePupxlQX0F2R7UNTerNP0Z3e1JNGOMw8/view?usp=sharing'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
top10_address = pd.read_csv(path)

In [ ]:
top10_address

street  ...   latitude
0               город Москва, Ходынский бульвар, дом 4  ...  55.790231
1          город Москва, Пресненская набережная, дом 2  ...  55.749162
2       город Москва, проспект Мира, дом 211, корпус 2  ...  55.845855
3          город Москва, Кировоградская улица, дом 13А  ...  55.612182
4           город Москва, Кутузовский проспект, дом 57  ...  55.727932
..                                                 ...  ...        ...
520  город Москва, Ленинградский проспект, дом 33, ...  ...  55.786460
521              город Москва, проспект Мира, дом 114Б  ...  55.808154
522  город Москва, Профсоюзная улица, дом 25, корпус 1  ...  55.674006
523             город Москва, Профсоюзная улица, дом 9  ...  55.682422
524            город Москва, Варшавское шоссе, дом 206  ...  55.537088

[525 rows x 4 columns]

С помощью folium визуализируем адреса заведений по нашим топ 10 улицам

In [ ]:
moscow = folium.Map(location = [55.755819, 37.617644], zoom_start = 11)

coordinates = []

latitudes = list(top10_address['latitude'])
longitudes = list(top10_address['longitude'])

for lat, lng in zip(latitudes, longitudes):
    coordinates.append([lat, lng])

moscow.add_child(plugins.HeatMap(coordinates))

### Найдем число улиц с одним объектом общественного питания. Воспользуемся внешней информацией для ответа на вопрос — в каких районах Москвы находятся эти улицы?


In [ ]:
one_obj_street = (
    rest_data
        .groupby('street', as_index=False)
        .agg({'id':'count'})
        .query('id == 1')
        .drop(columns='id')
        .reset_index(drop=True)
)

one_obj_street

street
0      1-й Балтийский переулок
1       1-й Басманный переулок
2      1-й Ботанический проезд
3      1-й Вешняковский проезд
4    1-й Голутвинский переулок
..                         ...
596             улица Шумилова
597              улица Шумкина
598                улица Щорса
599             улица Энгельса
600               шоссе Фрезер

[601 rows x 1 columns]

**Пропускаем**

In [ ]:
def get_location_yandexgeo(row):
    street = '+'.join(row['street'].split())

    url = 'https://geocode-maps.yandex.ru/1.x'
    params = {'geocode':'Москва,{}'.format(street),
              'format':'json',
              'apikey': API}

    r = requests.get(url, params=params).json()

    try:
        for i in r['response']['GeoObjectCollection']['featureMember'][0]['GeoObject']['metaDataProperty']['GeocoderMetaData']['Address']['Components']:
            if 'округ' in i['name']:
                region = i['name']  
    except:
        region = ''      

    try:
        pos = r['response']['GeoObjectCollection']['featureMember'][0]['GeoObject']['Point']['pos']
        return pd.Series([float(pos.split()[0]), float(pos.split()[1]), region])
    except:
        return pd.Series([0, 0, region])

In [ ]:
tqdm.pandas()

one_obj_street[['longitude', 'latitude', 'region']] = one_obj_street.progress_apply(get_location_yandexgeo, axis=1)

Теперь, так как район можно получить только при обратном геокодировании, мы еще раз напишем функцию, которая будет нам искать район

*У большинства улиц данные о районе находятся вот здесь ['response']['GeoObjectCollection']['featureMember'][1], но есть и такие ['response']['GeoObjectCollection']['featureMember'][0], поэтому в функции такой дикий цикл прописан))*

In [ ]:
def get_district_yandexgeo(row):
    url = 'https://geocode-maps.yandex.ru/1.x'
    params = {'geocode':'{},{}'.format(row['longitude'], row['latitude']),
              'format':'json',
              'apikey': API}

    r = requests.get(url, params=params).json()          
  
    try:
        for i in r['response']['GeoObjectCollection']['featureMember'][1]['GeoObject']['metaDataProperty']['GeocoderMetaData']['Address']['Components']:
            if 'район' in i['name']:
                return i['name']
        else:
            for j in r['response']['GeoObjectCollection']['featureMember'][0]['GeoObject']['metaDataProperty']['GeocoderMetaData']['Address']['Components']:
                if 'район' in j['name']:
                    return j['name']
    except:
        return ''

In [ ]:
tqdm.pandas()

one_obj_street['district'] = one_obj_street.progress_apply(get_district_yandexgeo, axis=1)

/usr/local/lib/python3.7/dist-packages/tqdm/std.py:658: FutureWarning:

The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version



Лимит запросов в сутки 1000, но у меня выходит 1202...

![image.jpg](https://www.meme-arsenal.com/memes/00fef54d01d5d0393c437f94c49e76a1.jpg)

In [ ]:
one_obj_street.to_csv('one_obj_street.csv', encoding='utf-8', index=False)

**Вот отсюда можно запускать**

In [ ]:
url = 'https://drive.google.com/file/d/1CSAtONviCnPMUp9RdC426HQe-fSR8fCr/view?usp=sharing'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
one_obj_street = pd.read_csv(path)

In [ ]:
one_obj_street.head()

street  ...         district
0    1-й Балтийский переулок  ...   район Аэропорт
1     1-й Басманный переулок  ...  Басманный район
2    1-й Ботанический проезд  ...   район Свиблово
3    1-й Вешняковский проезд  ...  Рязанский район
4  1-й Голутвинский переулок  ...   район Якиманка

[5 rows x 5 columns]

Теперь проверим хорошо ли всё подтянулось

Проверим, есть ли пустые значения

In [ ]:
one_obj_street.query('longitude == 0')

street  longitude  latitude region district
500  улица 1-й километр Объездной Дороги        0.0       0.0    NaN      NaN

1 строка, заполню вручную

In [ ]:
one_obj_street.loc[one_obj_street['latitude'] == 0, ['longitude', 'latitude', 'region']] = 37.350458, 55.328965, 'Троицкий административный округ'

In [ ]:
one_obj_street.query('region != region')

Empty DataFrame
Columns: [street, longitude, latitude, region, district]
Index: []

In [ ]:
one_obj_street.query('district != district')

street  ...  district
5                      1-й Заречный переулок  ...       NaN
31                     1-я Ватутинская улица  ...       NaN
38   1-я линия Хорошёвского Серебряного Бора  ...       NaN
54                   2-й Магистральный тупик  ...       NaN
66                     2-я Барышевская улица  ...       NaN
..                                       ...  ...       ...
572                   улица Потаповская Роща  ...       NaN
575           улица Рассудовское Лесничество  ...       NaN
587                            улица Слобода  ...       NaN
589                   улица Старые Кузьминки  ...       NaN
595                     улица Школа-интернат  ...       NaN

[64 rows x 5 columns]

64 пропуска - 10%. Вероятно, их можно было бы удалить, но я попробую заполнить их с помощью данных мос газа, где хранятся районы с привязкой к улице.

*Можно было бы сразу использовать мосгаз без геокодера, но в мосгазе есть небольшой недостаток, внизу я написал о нем (где дубликаты)*

In [ ]:
url = 'https://drive.google.com/file/d/1kX0nEv4knDQ4mVWhtVnbqgJqSoZLUouD/view?usp=sharing'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
mosgaz = pd.read_csv(path)

In [ ]:
mosgaz.head()

streetname  areaid okrug               area
0             Выставочный переулок      17   ЦАО  Пресненский район
1                     улица Гашека      17   ЦАО  Пресненский район
2          Большая Никитская улица      17   ЦАО  Пресненский район
3                Глубокий переулок      17   ЦАО  Пресненский район
4  Большой Гнездниковский переулок      17   ЦАО  Пресненский район

Мерджим таблицы

In [ ]:
one_obj_street = pd.merge(one_obj_street, mosgaz[['streetname', 'area']], left_on='street', right_on='streetname', how='left')
one_obj_street.drop(columns='streetname', inplace=True)
one_obj_street

street  ...                  area
0    1-й Балтийский переулок  ...        Район Аэропорт
1     1-й Басманный переулок  ...       Басманный район
2     1-й Басманный переулок  ...  Красносельский район
3    1-й Ботанический проезд  ...        Район Свиблово
4    1-й Вешняковский проезд  ...       Рязанский район
..                       ...  ...                   ...
643           улица Шумилова  ...       Район Кузьминки
644            улица Шумкина  ...      Район Сокольники
645              улица Щорса  ...        Район Солнцево
646           улица Энгельса  ...                   NaN
647             шоссе Фрезер  ...   Нижегородский район

[648 rows x 6 columns]

Проверим, есть ли улицы, которых и мосгаз не спасет

In [ ]:
print(one_obj_street.query('district != district and area != area').shape)
one_obj_street.query('district != district and area != area')

(37, 6)


street  longitude  ...  district area
6                  1-й Заречный переулок  36.958128  ...       NaN  NaN
33                 1-я Ватутинская улица  37.326357  ...       NaN  NaN
70                 2-я Барышевская улица  37.543911  ...       NaN  NaN
107       39-й километр Калужского шоссе  37.319889  ...       NaN  NaN
145                      Армейская улица  37.371164  ...       NaN  NaN
448                     Физическая улица  37.292733  ...       NaN  NaN
450                    Фитарёвская улица  37.480606  ...       NaN  NaN
465                      Цветочная улица  37.131988  ...       NaN  NaN
490                      деревня Бабенки  37.187198  ...       NaN  NaN
491                      деревня Бунчиха  37.036856  ...       NaN  NaN
492              деревня Верхнее Валуево  37.355121  ...       NaN  NaN
493                      деревня Внуково  37.266358  ...       NaN  NaN
494                 деревня Голохвастово  37.202461  ...       NaN  NaN
495                        деревня Десна  37.370967  ...       NaN  NaN
496                      деревня Каменка  37.137324  ...       NaN  NaN
497                       деревня Ликова  37.274775  ...       NaN  NaN
498                      деревня Пыхтино  37.298760  ...       NaN  NaN
499            деревня Сенькино-Секерино  37.138887  ...       NaN  NaN
500                     деревня Столбово  37.469808  ...       NaN  NaN
501                     деревня Шаганино  37.372575  ...       NaN  NaN
502                  деревня Яковлевское  36.940386  ...       NaN  NaN
503                       деревня Ясенки  37.214894  ...       NaN  NaN
504                         квартал № 30  37.607080  ...       NaN  NaN
511                          корпус 4Г-Д  36.867344  ...       NaN  NaN
524                   посёлок Газопровод  37.475513  ...       NaN  NaN
525                    посёлок Остафьево  37.485816  ...       NaN  NaN
526                       посёлок Рогово  37.076292  ...       NaN  NaN
535                          село Былово  37.243972  ...       NaN  NaN
536     территория ТСЖ "Лесная поляна-2"  37.465488  ...       NaN  NaN
537  улица 1-й километр Объездной Дороги  37.350458  ...       NaN  NaN
569              улица Дмитрия Рябинкина  37.331872  ...       NaN  NaN
574              улица Евгения Родионова  37.343514  ...       NaN  NaN
615               улица Потаповская Роща  37.494647  ...       NaN  NaN
618       улица Рассудовское Лесничество  36.931421  ...       NaN  NaN
631                        улица Слобода  37.660341  ...       NaN  NaN
634               улица Старые Кузьминки  37.790839  ...       NaN  NaN
642                 улица Школа-интернат  37.297134  ...       NaN  NaN

[37 rows x 6 columns]

37 пропусков. Можно предположить, что эти пропуски у нас из-за того, что данные мосгаза за 2013 год. Тогда эти улицы считались Московской областью, а сейчас стали частью Москвы. Почти все пропуски находятся в Троицком и Новомосковском административных округах. С ними я уже ничего не сделаю, поэтому удалю их

In [ ]:
die = list(one_obj_street.query('district != district and area != area').index)
die

[6,
 33,
 70,
 107,
 145,
 448,
 450,
 465,
 490,
 491,
 492,
 493,
 494,
 495,
 496,
 497,
 498,
 499,
 500,
 501,
 502,
 503,
 504,
 511,
 524,
 525,
 526,
 535,
 536,
 537,
 569,
 574,
 615,
 618,
 631,
 634,
 642]

In [ ]:
one_obj_street = one_obj_street.query('index not in @die')
one_obj_street

street  ...                  area
0    1-й Балтийский переулок  ...        Район Аэропорт
1     1-й Басманный переулок  ...       Басманный район
2     1-й Басманный переулок  ...  Красносельский район
3    1-й Ботанический проезд  ...        Район Свиблово
4    1-й Вешняковский проезд  ...       Рязанский район
..                       ...  ...                   ...
643           улица Шумилова  ...       Район Кузьминки
644            улица Шумкина  ...      Район Сокольники
645              улица Щорса  ...        Район Солнцево
646           улица Энгельса  ...                   NaN
647             шоссе Фрезер  ...   Нижегородский район

[611 rows x 6 columns]

In [ ]:
print(one_obj_street.query('district != district').shape)
one_obj_street.query('district != district')

(33, 6)


street  ...                        area
40   1-я линия Хорошёвского Серебряного Бора  ...     Район Хорошево-Мневники
58                   2-й Магистральный тупик  ...           Хорошевский район
125                        6-я улица Лазенки  ...      Район Ново-Переделкино
141                    Андроньевская площадь  ...             Таганский район
149                       Аэрофлотская улица  ...               Район Внуково
164                 Большая Внуковская улица  ...               Район Внуково
180       Большой Трёхсвятительский переулок  ...             Басманный район
219                        Дивизионная улица  ...               Район Внуково
227                             Жуков проезд  ...         Район Замоскворечье
228                             Жуков проезд  ...           Даниловский район
234                       Иваньковское шоссе  ...  Район Покровское-Стрешнево
242                  Интернациональная улица  ...               Район Внуково
261                        Кузьминская улица  ...             Район Кузьминки
262                        Кузьминская улица  ...       Район Выхино-Жулебино
285                Малый Ивановский переулок  ...             Басманный район
319                        Народный проспект  ...             Район Измайлово
330                      Новолучанская улица  ...               Район Кунцево
333                     Новорублёвская улица  ...               Район Кунцево
342                           Парковая улица  ...      Район Ново-Переделкино
343                           Парковая улица  ...                Район Митино
352                  Петроверигский переулок  ...             Басманный район
361                            Полевая улица  ...               Район Внуково
362                            Полевая улица  ...               Район Крюково
363                            Полевая улица  ...          Район Южное Бутово
398                     Софийская набережная  ...              Район Якиманка
427                          Тенистый проезд  ...              Район Свиблово
435                     Триумфальная площадь  ...              Тверской район
439                     Тургеневская площадь  ...        Красносельский район
459                      Хитровский переулок  ...             Басманный район
464                      Хохловский переулок  ...             Басманный район
488                          Яузский бульвар  ...             Таганский район
521                        площадь Журавлёва  ...        Район Преображенское
522                        площадь Журавлёва  ...        Район Соколиная Гора

[33 rows x 6 columns]

33 пропуска можем заполнить

In [ ]:
one_obj_street['district'].fillna(one_obj_street['area'], inplace=True)

In [ ]:
one_obj_street.query('district != district')

Empty DataFrame
Columns: [street, longitude, latitude, region, district, area]
Index: []

In [ ]:
one_obj_street.shape

(611, 6)

Проверим дубликаты

In [ ]:
one_obj_street['street'].duplicated().sum()

47

Так как у нас только улицы (без номера дома), то в случаях, когда одна улица находится в нескольких районах, мы получаем дубликаты. Я считаю, это не особо важно, потому что районы находятся рядом, поэтому дубликаты можно удалить и оставить 1 любой район

In [ ]:
one_obj_street.drop_duplicates(subset='street', inplace=True)

In [ ]:
one_obj_street.shape

(564, 6)

Ну вот, почти половину пропусков получилось заполнить.

И удалим ненужный столбец

In [ ]:
one_obj_street.drop(columns='area', inplace=True)
one_obj_street

street  ...             district
0      1-й Балтийский переулок  ...       район Аэропорт
1       1-й Басманный переулок  ...      Басманный район
3      1-й Ботанический проезд  ...       район Свиблово
4      1-й Вешняковский проезд  ...      Рязанский район
5    1-й Голутвинский переулок  ...       район Якиманка
..                         ...  ...                  ...
643             улица Шумилова  ...      район Кузьминки
644              улица Шумкина  ...     район Сокольники
645                улица Щорса  ...       район Солнцево
646             улица Энгельса  ...      Басманный район
647               шоссе Фрезер  ...  Нижегородский район

[564 rows x 5 columns]

In [ ]:
pd.DataFrame(round((one_obj_street.isna().mean()*100),2)).style.background_gradient('coolwarm')

0 пропусков - отлично

Наконец можем нарисовать тепловую карту

In [ ]:
moscow = folium.Map(location = [55.755819, 37.617644], zoom_start = 11)

coordinates = []

latitudes = list(one_obj_street['latitude'])
longitudes = list(one_obj_street['longitude'])

for lat, lng in zip(latitudes, longitudes):
    coordinates.append([lat, lng])

moscow.add_child(plugins.HeatMap(coordinates))

Построим график топ-10 районов

In [ ]:
dr = (
    one_obj_street
        .groupby('district', as_index=False)
        .agg({'street':'count'})
        .sort_values(by='street')
        .rename(columns={'street':'count'})
)

fig = px.bar(dr.tail(10), x='count', y='district', orientation='h',
             title='Топ-10 районов Москвы по количеству улиц с 1 заведением')
fig.show()

Тверской и таганский районы лидируют по количеству улиц с 1 заведением - 28 и 27 улиц соответственно.

Нагляднее будет на карте

In [ ]:
top10_district_alone = dr.tail(10).reset_index(drop=True)
top10_district_alone

district  count
0      район Лефортово     12
1  район Замоскворечье     14
2      Мещанский район     16
3   район Марьина Роща     17
4    Пресненский район     18
5  Тимирязевский район     19
6      район Хамовники     23
7      Басманный район     25
8      Таганский район     27
9       Тверской район     28

**не запускать**

In [ ]:
def get_location_yandexgeo1(row):
    district = '+'.join(row['district'].split())

    url = 'https://geocode-maps.yandex.ru/1.x'
    params = {'geocode':'Москва,{}'.format(district),
              'format':'json',
              'apikey': API}

    r = requests.get(url, params=params).json()     

    try:
        pos = r['response']['GeoObjectCollection']['featureMember'][0]['GeoObject']['Point']['pos']
        return pd.Series([float(pos.split()[0]), float(pos.split()[1])])
    except:
        return pd.Series([0, 0])

In [ ]:
tqdm.pandas()

top10_district_alone[['longitude', 'latitude']] = top10_district_alone.progress_apply(get_location_yandexgeo1, axis=1)

/usr/local/lib/python3.7/dist-packages/tqdm/std.py:658: FutureWarning:

The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version



In [ ]:
top10_district_alone.to_csv('top10_district_alone.csv', encoding='utf-8', index=False)

**Отсюда можно запускать**

In [ ]:
url = 'https://drive.google.com/file/d/1yZMcHCkZ4Q4yWIjE_ZeQ0ATQJGvH58xt/view?usp=sharing'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
top10_district_alone = pd.read_csv(path)

In [ ]:
top10_district_alone

district  count  longitude   latitude
0      район Лефортово     12  37.704826  55.753635
1  район Замоскворечье     14  37.634290  55.734162
2      Мещанский район     16  37.628801  55.780071
3   район Марьина Роща     17  37.614383  55.796936
4    Пресненский район     18  37.562389  55.763437
5  Тимирязевский район     19  37.557071  55.825527
6      район Хамовники     23  37.574534  55.729199
7      Басманный район     25  37.671238  55.766572
8      Таганский район     27  37.666971  55.740010
9       Тверской район     28  37.606918  55.770132

In [ ]:
moscow = folium.Map(location = [55.755819, 37.617644], zoom_start = 11)

for indice, row in top10_district_alone.iterrows():
    folium.Marker(
        location=[row['latitude'], row['longitude']],
        popup=row['district']
    ).add_to(moscow)

moscow

*Смотреть по округам смысла нет, т.к. там практически все заведения находятся в центральном округе*

Так же, так как наше заведение будет кафе, то отобразим на карте все кафе, что поможет нам выбрать хорошее место

In [ ]:
cafe_data = (
    rest_data
        .query('object_type == "кафе"')
        .groupby('street', as_index=False)
        .agg({'id':'count'})
        .sort_values(by='id', ascending=False)
)

cafe_data

street   id
944             проспект Мира  103
625    Пресненская набережная   98
632         Профсоюзная улица   87
428    Ленинградский проспект   70
430        Ленинский проспект   64
..                        ...  ...
99    3-я Подрезковская улица    1
710           Солнечная аллея    1
715            Сосновая улица    1
716      Софийская набережная    1
0    1-й Автозаводский проезд    1

[1279 rows x 2 columns]

Координаты многих улиц у нас уже есть в датафрейме one_obj_street

In [ ]:
one_obj_street

street  ...             district
0      1-й Балтийский переулок  ...       район Аэропорт
1       1-й Басманный переулок  ...      Басманный район
3      1-й Ботанический проезд  ...       район Свиблово
4      1-й Вешняковский проезд  ...      Рязанский район
5    1-й Голутвинский переулок  ...       район Якиманка
..                         ...  ...                  ...
643             улица Шумилова  ...      район Кузьминки
644              улица Шумкина  ...     район Сокольники
645                улица Щорса  ...       район Солнцево
646             улица Энгельса  ...      Басманный район
647               шоссе Фрезер  ...  Нижегородский район

[564 rows x 5 columns]

In [ ]:
cafe_data = pd.merge(cafe_data, one_obj_street[['street', 'longitude', 'latitude']], on='street', how='left')
cafe_data

street   id  longitude   latitude
0                проспект Мира  103        NaN        NaN
1       Пресненская набережная   98        NaN        NaN
2            Профсоюзная улица   87        NaN        NaN
3       Ленинградский проспект   70        NaN        NaN
4           Ленинский проспект   64        NaN        NaN
...                        ...  ...        ...        ...
1274   3-я Подрезковская улица    1  37.353701  55.940961
1275           Солнечная аллея    1  37.194717  55.980709
1276            Сосновая улица    1        NaN        NaN
1277      Софийская набережная    1  37.617977  55.747875
1278  1-й Автозаводский проезд    1        NaN        NaN

[1279 rows x 4 columns]

In [ ]:
cafe_data.query('longitude == longitude')

street  id  longitude   latitude
750          1-й Митинский переулок   1  37.378692  55.845446
756                 проезд Досфлота   1  37.457834  55.845567
757   2-й Новоподмосковный переулок   1  37.509757  55.818114
764              1-я Боевская улица   1  37.693282  55.791830
765           улица Полины Осипенко   1  37.532825  55.779216
...                             ...  ..        ...        ...
1269               Спортивная улица   1  37.668471  55.609736
1270             3-я Северная линия   1  37.552534  55.942368
1274        3-я Подрезковская улица   1  37.353701  55.940961
1275                Солнечная аллея   1  37.194717  55.980709
1277           Софийская набережная   1  37.617977  55.747875

[161 rows x 4 columns]

Ага, оказывается не так уж и много... 

Остальное тогда вытащу, используя api геокодера

**запускать не нужно**

In [ ]:
tqdm.pandas()

cafe_data[['longitude', 'latitude']] = cafe_data.query('longitude != longitude').progress_apply(get_location_yandexgeo_2, axis=1)

In [ ]:
cafe_data.to_csv('cafe_data.csv', encoding='utf-8', index=False)

**вот отсюда**

In [ ]:
url = 'https://drive.google.com/file/d/18Hd2xlbBEkZH4cxXIAJJwNUVUrSJrBXp/view?usp=sharing'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
cafe_data = pd.read_csv(path)

In [ ]:
cafe_data

street   id  longitude   latitude
0                проспект Мира  103  37.637937  55.812368
1       Пресненская набережная   98  37.540991  55.746791
2            Профсоюзная улица   87  37.532511  55.649525
3       Ленинградский проспект   70  37.545635  55.794285
4           Ленинский проспект   64  37.537146  55.682630
...                        ...  ...        ...        ...
1274         Фитарёвская улица    1  37.480606  55.567107
1275        Коптевский бульвар    1        NaN        NaN
1276            Флотская улица    1  37.498267  55.853026
1277        Конюшковская улица    1  37.577427  55.755996
1278  1-й Автозаводский проезд    1  37.656631  55.704915

[1279 rows x 4 columns]

Функция значит заменила на NaN те 160 строк, что мы заполнили мерджем, так что опять замерджим

In [ ]:
cafe_data = pd.merge(cafe_data, one_obj_street[['street', 'longitude', 'latitude']], on='street', how='left')
cafe_data

street   id  ...  longitude_y  latitude_y
0                проспект Мира  103  ...          NaN         NaN
1       Пресненская набережная   98  ...          NaN         NaN
2            Профсоюзная улица   87  ...          NaN         NaN
3       Ленинградский проспект   70  ...          NaN         NaN
4           Ленинский проспект   64  ...          NaN         NaN
...                        ...  ...  ...          ...         ...
1274         Фитарёвская улица    1  ...          NaN         NaN
1275        Коптевский бульвар    1  ...    37.525154   55.831069
1276            Флотская улица    1  ...          NaN         NaN
1277        Конюшковская улица    1  ...          NaN         NaN
1278  1-й Автозаводский проезд    1  ...          NaN         NaN

[1279 rows x 6 columns]

In [ ]:
cafe_data = (
    cafe_data
        .fillna(
            {'longitude_x':cafe_data['longitude_y'],
             'latitude_x':cafe_data['latitude_y']})
        .drop(
            columns=['longitude_y', 'latitude_y'])
        .rename(columns={'longitude_x':'longitude',
                         'latitude_x':'latitude'})
)

cafe_data

Датафрейм готов

In [ ]:
moscow = folium.Map(location = [55.755819, 37.617644], zoom_start = 11)

coordinates = []

latitudes = list(cafe_data['latitude'])
longitudes = list(cafe_data['longitude'])

for lat, lng in zip(latitudes, longitudes):
    coordinates.append([lat, lng])

moscow.add_child(plugins.HeatMap(coordinates))

### Посмотрим на распределение количества посадочных мест для улиц с большим количеством объектов общественного питания. Какие закономерности можно выявить?


In [ ]:
number_top10 = (
    rest_data
        .query('street in @top10_street.street')
        .groupby('street')
        ['number'].agg(['count', 'sum', 'mean'])
)

number_top10['mean'] = (np.ceil(number_top10['mean'])).astype('int')
number_top10

count    sum  mean
street                                    
Варшавское шоссе          160   8456    53
Кировоградская улица      107   6409    60
Кутузовский проспект      114   9697    86
Ленинградский проспект    170   8970    53
Ленинский проспект        146   9080    63
Пресненская набережная    167   7656    46
Профсоюзная улица         178   8355    47
Ходынский бульвар         102   1933    19
проспект Вернадского      126   8612    69
проспект Мира             204  12790    63

In [ ]:
fig = px.bar(number_top10, orientation='h',
             y=number_top10.index, x='mean',
             title='Распределение посадочных мест для улиц с большим количеством заведений')

fig.update_layout(
    xaxis_title='number'
)

fig.show()

In [ ]:
int(np.ceil(number_top10['mean'].mean()))

56

Для улиц с большим количество объектов характерны заведения с большим количеством посадочных мест (в среднем по 56 мест).

# Общий вывод и рекомендации

1. 40% всех заведений Москвы - это кафе. Сильная конкуренция для нас. Рестораны занимают 15%. В принципе можно и ресторан открыть, а не кафе.
2. У кафе и ресторанов примерно одинаковое количество сетевых заведений - в районе 23%. Для нас несомненный плюс - сетевые заведения довольно серьезные конкуренты, и могли бы спокойно выдавить нас с рынка на начальном этапе.
3. Лидер по сетевым заведениям - предприятия быстрого обслуживания. По сути они не наши прямые конкуренты, так что страшиться нам нечего.
4. В среднем в кафе по 30 посадочных мест, в ресторанах - 80. Для нас эти цифры будет наглядным ориентиром. Больше 80 посадочных мест делать не нужно, но и меньше 30 определенно точно не стоит. 50 вполне приемлемо.
5. Судя по тепловой карте топ-10 улиц по количеству заведений, кафе/ресторан нам открывать лучше на Юго-Востоке Москвы. Меньше заведений - меньше конкуренция.

Итак, открываем кафе/ресторан на Юго-Востоке Москвы на 50 посадочных мест.



Презентация: https://disk.yandex.ru/i/cy-t_LTgvWbFAg